base_layer/wallet/src/storage/database.rs

```diff
@@ -0,0 +1,226 @@
+// Copyright 2019. The Tari Project
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::error::WalletStorageError;
+use log::*;
+use std::fmt::{Display, Error, Formatter};
+use tari_comms::{peer_manager::Peer, types::CommsPublicKey};
+
+const LOG_TARGET: &'static str = "wallet::contacts_service::database";
+
+/// This trait defines the functionality that a database backend need to provide for the Contacts Service
+pub trait WalletBackend: Send + Sync {
+    /// Retrieve the record associated with the provided DbKey
+    fn fetch(&self, key: &DbKey) -> Result<Option<DbValue>, WalletStorageError>;
+    /// Modify the state the of the backend with a write operation
+    fn write(&mut self, op: WriteOperation) -> Result<Option<DbValue>, WalletStorageError>;
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum DbKey {
+    Peer(CommsPublicKey),
+    Peers,
+}
+
+pub enum DbValue {
+    Peer(Box<Peer>),
+    Peers(Box<Vec<Peer>>),
```

Comment:
Vec is already on the heap, so you probably don't need to Box here

Reply:
kk